### ISSUE TITLE
Hive: Support dynamic partition insert

### ISSUE THREAD

**Issue by barunpuri** [sqlfluff/sqlfluff#3117]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff currently handle only for static partition insert. So if sql uses dynamic partition insert, this makes error.

### Expected Behaviour

sqlfluff supports to parse dynamic partition insert

### Observed Behaviour

hql with dynamic partition insert
```
INSERT OVERWRITE TABLE aaa.bbb PARTITION (stats_id, log_date)
SELECT
    'a' as stats_id,
    'b' as log_date
FROM ccc.ddd
```

sqlfluff returns
```
 ~ >> sqlfluff lint de-scripts/test.hql                                                                
== [de-scripts/test.hql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'INSERT
                       | OVERWRITE TABLE aaa.bbb PARTITION...'
All Finished 📜 🎉!
```

hql with static partition insert
```
INSERT OVERWRITE TABLE aaa.bbb PARTITION (stats_id = 'a', log_date = 'b')
SELECT
    'a' AS a,
    'b' AS b
FROM ccc.ddd
```

sqlfluff returns
```
 ~ >> sqlfluff lint de-scripts/test.hql                                                                
All Finished 📜 🎉!
```

### How to reproduce

Add some sequence into `PartitionSpecGrammer` in `dialect_hive.py `
https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_hive.py#L123-L135

### Dialect

hive

### Version

sqlfluff, version 0.12.0
python 3.7.10

### Configuration

[sqlfluff]
dialect = hive
sql_file_exts = .sql,.hql
exclude_rules = L025,L031
templater = raw

[sqlfluff:indentation]
indented_using_on = true
template_blocks_indent = false

[sqlfluff:templater:jinja:context]
a=a

[sqlfluff:templater:python: context]
a=a

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 132
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = unqualified
unquoted_identifiers_policy = all

[sqlfluff:rules:L010]
capitalisation_policy = upper

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower
ignore_words = None

[sqlfluff:rules:L016]
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:L022]
comma_style = trailing

[sqlfluff:rules:L030]
extended_capitalisation_policy = upper
ignore_words = None

[sqlfluff:rules:L031]
force_enable = False

[sqlfluff:rules:L040]
capitalisation_policy = upper
ignore_words = None

[sqlfluff:rules:L042]
forbid_subquery_in = join

[sqlfluff:rules:L051]
fully_qualify_join_types = both

[sqlfluff:rules:L054]
group_by_and_order_by_style = explicit

[sqlfluff:rules:L063]
extended_capitalisation_policy = upper
ignore_words = None

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
This is due to this:

https://github.com/sqlfluff/sqlfluff/blob/41e2920f46860bb25455d456c222d6fa1e8a3ae6/src/sqlfluff/dialects/dialect_hive.py#L123-L134

You could make the Equals part optional if that's the case:

```py
    PartitionSpecGrammar=Sequence(
        "PARTITION",
        Bracketed(
            Delimited(
                Sequence(
                    Ref("ColumnReferenceSegment"),
                    Sequence(
                        Ref("EqualsSegment"),
                        Ref("LiteralGrammar"),
                        optional=True
                    ),
                )
            )
        ),
    ),
```

**Comment by tunetheweb** [comment]
So that would mean either all have to have an Equals Segment or none of them do?

Your suggestion is almost right, but then won't enforce a Delimiter between each one so don't think you need the `AnyNumberOf`.

However, I find the following clearer, as don't have to realise/remember that `Delimited` inherits from `OneOf`:

```py
    PartitionSpecGrammar=Sequence(
        "PARTITION",
        Bracketed(
            OneOf(
                Delimited(
                    Sequence(
                        Ref("ColumnReferenceSegment"),
                        Ref("EqualsSegment"),
                        Ref("LiteralGrammar"),
                    )
                ),
                Delimited(
                    Ref("ColumnReferenceSegment"),
                )
        ),
    ),
```

**Comment by barunpuri** [comment]
Sorry for make you confuse. 
It was meant
- Equal Segment(static) &rarr; Equal Segment : permitted
- Equal Segment &rarr; Non Equal Segment : permitted
- Non Equal Segment(dynamic) &rarr; Non Equal Segment : permitted
- Non Equal Segment &rarr; Equal Segment : Not permitted

**Comment by tunetheweb** [comment]
Sorry I don't understand what you mean by "Equal Segment(static)" and "Equal Segment(dynamic)"

Could you give SQL examples for those 4 cases?

**Comment by barunpuri** [comment]
Case 1 : Equal Segment &rarr; Equal Segment : permitted
```
INSERT OVERWRITE TABLE foo
PARTITION (a = 'test_foo', b = 'test_bar') IF NOT EXISTS
SELECT a, b FROM bar;
```

Case 2 : Equal Segment &rarr; None Equal Segment : permitted
```
INSERT OVERWRITE TABLE foo
PARTITION (a = 'test_foo', b) IF NOT EXISTS
SELECT a, 'test_bar' AS b FROM bar;
```

Case 3 : None Equal Segment &rarr; None Equal Segment : permitted
```
INSERT OVERWRITE TABLE foo
PARTITION (a, b) IF NOT EXISTS
SELECT 'test_bar' AS a, 'test_bar' AS b FROM bar;
```

Case 4 : None Equal Segment &rarr; Equal Segment : Not permitted
```
INSERT OVERWRITE TABLE foo
PARTITION (a, b = 'test_bar') IF NOT EXISTS
SELECT 'test_foo' AS a, b FROM bar;
```

**Comment by barunpuri** [comment]
All right, I understand the purpose of this project and I accept your suggestion. https://github.com/sqlfluff/sqlfluff/issues/3117#issuecomment-1102660112 
I have one more question. 
Can I write a pr that merges into master with modifying the code in the way you suggested?

**Comment by barunpuri** [comment]
I made only test case 2. 
Because test case 1 exists on test cases and I think test case 3 is included in 2
